BUG: fix uint64 overflow in triu_indices using operator.index

Kairoven opened the pull request:
Fixes #29488 

This PR fixes an OverflowError and RuntimeWarning when using `np.uint64` inputs with `np.triu_indices`(#29488).

Following a suggestion from @seberg, this PR ensures robust handling of unsigned integer inputs in `triu_indices`.

Initially, only `n` was converted using `operator.index(n)`. However, testing revealed that this partial fix still led to `RuntimeWarning: overflow encountered in scalar subtract`.

So I later added `operator.index()` for `m` and `k` as well to ensure all inputs are safely normalized:
- `n = operator.index(n)`
- `m = operator.index(m)` if provided
- `k = operator.index(k)`

This ensures consistent and safe integer handling across all inputs, preventing overflow and aligning behavior with functions like `tril_indices`.

A new test `test_triu_indices_with_uint64` has been added to verify correct behavior with `uint64` inputs.

Kairoven committed:
BUG: fix overflow in triu_indices with uint64 input

Normalize n using operator.index to handle unsigned integer inputs.
This fixes an OverflowError when k-1=-1 is used in tri() with uint64 shape.
Closes #29488

Kairoven committed:
BUG: fix uint64 overflow in triu_indices with proper input normalization

Normalize n, m, k using operator.index to handle unsigned integer inputs.
This fixes an OverflowError when k-1=-1 is used in tri() with uint64 shape.
Closes #29488

Kairoven committed:
TEST: add test for triu_indices with uint64 input

Ensure that triu_indices works correctly when n, m, or k are np.uint64.
The function should not raise OverflowError or RuntimeWarning,
and should produce the same result as with int inputs.

Kairoven committed:
FIX: remove whitespace in blank lines (linter fix)

Kairoven committed:
FIX: add self and remove whitespace in test_triu_indices_with_uint64

tylerjereddy commented on numpy/lib/tests/test_twodim_base.py:
```diff
@@ -510,6 +510,24 @@ def test_triu_indices(self):
+        # Compare with int version
+        iu_int = np.triu_indices(int(n), k=int(k), m=int(m))
+        assert_equal(iu[0], iu_int[0])
+        assert_equal(iu[1], iu_int[1])
```
(I somewhat wonder if it is slightly suboptimal to use the same function to produce the expected value; not sure NumPy regulars would care enough to want the expected value hard coded--I'll let them comment on that)

seberg reviewed (commented):
Two comments, I tend to agree with Tyler that it is nicer to spell it out.
The `m`/`n` handling would be correct in `tri`, but I somewhat hope it's not needed either way.

Kairoven commented on numpy/lib/_twodim_base_impl.py:
```diff
@@ -1132,6 +1132,13 @@ def triu_indices(n, k=0, m=None):
+    if m is None:
+        m = n
+    else:
+        m = operator.index(m)
```
Thank you so much for your feedback and for taking the time to review this — I truly appreciate your support.

After looking deeper into the issue, I now understand that problems can occur in the line:
`arange(-k, M - k, dtype=_min_int(-k, M - k))`
when `k` or `M` are unsigned integers (like `uint64`), because operations such as `-k` or `M - k` may cause overflow or underflow.

My current thinking is this: since `M` may come from `N` (when `M` is None), it would make sense to convert both `N` and `M` to Python integers using `operator.index()` at the start of the tri function, like this:
```
N = operator.index(N)
if like is not None:
    return _tri_with_like(like, N, M=M, k=k, dtype=dtype)

if M is None:
    M = N
else:
    M = operator.index(M)
```

So my plan is:

- Handle `N` and `M` in `tri`,

- Handle `k` in `triu_indices` (because it's often adjusted, e.g., k-1, and needs to be safe before being passed down).

But I’m not entirely sure if this is the cleanest or most consistent approach. I’d like to submit an updated commit. Would that be okay?

Kairoven commented:
After further investigation, I believe we may need to address this at the `tri` function level. Even if we fix `triu_indices`, similar issues will arise in functions like `tril_indices` when `n` and `m` are unsigned integers (e.g., `uint64`) and `k = -1`.

The root cause appears to be how NumPy handles arithmetic between `np.uint64` and Python integers. There's an inconsistency depending on whether the scalar is a NumPy integer or a native Python int. Here’s a minimal example:
```
n = np.uint64(2)
k1 = np.int64(-1)
k2 = -1

print(n + k1)  # outputs: 1.0 
print(n + k2)  # OverflowError: Python integer -1 out of bounds for uint64
```
In the `tri` function, this exact issue can occur in the line:
`arange(-k, M - k, dtype=_min_int(-k, M - k))`
When `k` is a Python int (like `-1`) and `M` is `uint64`, the subtraction `M - k` (or `-k`) can `raise an OverflowError`.

This makes `tri` — and by extension all functions that rely on it — fragile when handling mixed integer types.

I think one of the approach is to convert `N`, `M`, and `k` to safe Python integers using `operator.index()` at the very beginning of the tri function, before any arithmetic is performed.

Would you agree with this direction?

Kairoven committed:
Fix uint64 overflow in tri, triu(_indices), and tril_indices, ensure integer safety in tri.

In np.tri: N and M are now converted to Python integers using operator.index()  at the start, ensuring safe arithmetic in shape calculations.

In triu_indices and triu: the parameter k is also converted using operator.index() within each function.

Together, these changes ensure robust behavior when any of N, M, or k are  provided as unsigned integers, while maintaining correct semantics for all valid inputs.

Kairoven committed:
Add test for uint64 inputs in triu/tril_indices

Kairoven committed:
Fix formatting

Kairoven committed:
fix flake formatting

Kairoven committed:
Remove trailing whitespace

seberg reviewed (commented):
Thanks @Kairoven, seems like a safe thing to and looks now to be a minimal addition of `operator.index`.
Didn't think through if the tests cover everything, but they look reasonable
